# Análisis de datos de fútbol europeo: EDA y Dashboard

Proyecto final del máster en Data & Analytics. Análisis exploratorio de partidos de 
las cinco grandes ligas europeas (primera y segunda división), combinando dos fuentes 
de datos: resultados de partidos (Football-Data.co.uk) y puntuaciones Elo de los 
clubes (ClubElo).

## 1. Carga y reconocimiento de los datos

Cargo los dos conjuntos de datos en bruto y verifico sus dimensiones. Uso 
`low_memory=False` en el archivo de partidos para que pandas infiera correctamente 
los tipos de columna y evitar avisos de tipos mezclados.

In [252]:
import numpy as np
import pandas as pd

pd.set_option('display.max_columns', None)

df_matches = pd.read_csv('../data/raw/Matches.csv', low_memory=False)
df_elo = pd.read_csv('../data/raw/EloRatings.csv')

print("Partidos:", df_matches.shape)
print("Elo:", df_elo.shape)

Partidos: (230557, 48)
Elo: (245033, 4)


In [253]:
df_matches.head()

,Division,MatchDate,MatchTime,HomeTeam,AwayTeam,HomeElo,AwayElo,Form3Home,Form5Home,Form3Away,Form5Away,FTHome,FTAway,FTResult,HTHome,HTAway,HTResult,HomeShots,AwayShots,HomeTarget,AwayTarget,HomeFouls,AwayFouls,HomeCorners,AwayCorners,HomeYellow,AwayYellow,HomeRed,AwayRed,OddHome,OddDraw,OddAway,MaxHome,MaxDraw,MaxAway,Over25,Under25,MaxOver25,MaxUnder25,HandiSize,HandiHome,HandiAway,C_LTH,C_LTA,C_VHD,C_VAD,C_HTB,C_PHB
0,F1,2000-07-28,NaN,Marseille,Troyes,1686.34,1586.57,0.0,0.0,0.0,0.0,3.0,1.0,H,2.0,1.0,H,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.65,3.3,4.3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,F1,2000-07-28,NaN,Paris SG,Strasbourg,1714.89,1642.51,0.0,0.0,0.0,0.0,3.0,1.0,H,1.0,1.0,D,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.60,3.4,4.6,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,F2,2000-07-28,NaN,Wasquehal,Nancy,1465.08,1633.80,0.0,0.0,0.0,0.0,0.0,1.0,A,0.0,1.0,A,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,F1,2000-07-29,NaN,Auxerre,Sedan,1635.58,1624.22,0.0,0.0,0.0,0.0,0.0,1.0,A,0.0,1.0,A,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.70,3.3,3.9,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,F1,2000-07-29,NaN,Bordeaux,Metz,1734.34,1673.11,0.0,0.0,0.0,0.0,1.0,1.0,D,1.0,0.0,H,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.65,3.3,4.3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [254]:
df_elo.head()

,date,club,country,elo
0,2000-07-01,Aachen,GER,1453.60
1,2000-07-01,Aalborg,DEN,1482.61
2,2000-07-01,Aalst,BEL,1337.53
3,2000-07-01,Aarhus,DEN,1381.46
4,2000-07-01,Aberdeen,SCO,1360.43


In [255]:
df_matches.info(verbose=True, show_counts=True)

<class 'pandas.DataFrame'>
RangeIndex: 230557 entries, 0 to 230556
Data columns (total 48 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Division     230557 non-null  str    
 1   MatchDate    230557 non-null  str    
 2   MatchTime    99072 non-null   str    
 3   HomeTeam     230557 non-null  str    
 4   AwayTeam     230557 non-null  str    
 5   HomeElo      141597 non-null  float64
 6   AwayElo      141528 non-null  float64
 7   Form3Home    229057 non-null  float64
 8   Form5Home    229057 non-null  float64
 9   Form3Away    229057 non-null  float64
 10  Form5Away    229057 non-null  float64
 11  FTHome       230554 non-null  float64
 12  FTAway       230554 non-null  float64
 13  FTResult     230554 non-null  str    
 14  HTHome       175977 non-null  float64
 15  HTAway       175977 non-null  float64
 16  HTResult     175977 non-null  str    
 17  HomeShots    114735 non-null  float64
 18  AwayShots    114738 non-null  float

In [256]:
df_elo.info()


<class 'pandas.DataFrame'>
RangeIndex: 245033 entries, 0 to 245032
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   date     245033 non-null  str    
 1   club     245033 non-null  str    
 2   country  245033 non-null  str    
 3   elo      245033 non-null  float64
dtypes: float64(1), str(3)
memory usage: 7.5 MB


### Análisis de valores nulos

Antes de decidir qué columnas conservar, creo una función reutilizable que resume los 
valores nulos por columna, ordenados por porcentaje. Al ser una función, la aplico a 
las dos tablas sin duplicar código.

In [257]:
def resumen_nulos(df):
    """
    Genera un resumen de valores nulos por columna de un DataFrame.

    Parámetros:
    - df (pd.DataFrame): DataFrame a analizar.

    Retorna:
    - pd.DataFrame: tabla con el número de nulos y su porcentaje por columna,
      ordenada de mayor a menor porcentaje de nulos.
    """
    total_nulos = df.isnull().sum()
    porcentaje_nulos = (total_nulos / len(df)) * 100

    resumen = pd.DataFrame({
        "nulos": total_nulos,
        "porcentaje": porcentaje_nulos.round(2)
    })

    resumen_ordenado = resumen.sort_values(by="porcentaje", ascending=False)
    return resumen_ordenado

In [258]:
resumen_nulos(df_matches)

,nulos,porcentaje
MatchTime,131485,57.03
C_LTA,117955,51.16
C_VHD,117955,51.16
C_VAD,117955,51.16
C_LTH,117955,51.16
C_HTB,117955,51.16
C_PHB,117955,51.16
HomeTarget,116628,50.59
AwayTarget,116625,50.58
HomeFouls,116584,50.57


In [259]:
resumen_nulos(df_elo)

,nulos,porcentaje
date,0,0.0
club,0,0.0
country,0,0.0
elo,0,0.0


## 2. Acotación del alcance: las cinco grandes ligas

El análisis de nulos revela que muchas estadísticas detalladas solo están disponibles 
en las ligas principales. Analizo cuántos partidos hay por división para decidir el 
alcance del proyecto, garantizando superar el mínimo de 50.000 filas.

In [260]:
print("Número de ligas/divisiones distintas:", df_matches["Division"].nunique())
df_matches["Division"].value_counts()

Número de ligas/divisiones distintas: 38


Division
E1     13606
E2     13220
E3     12777
SP2    10497
I2      9921
E0      9410
I1      9012
SP1     9008
F1      8756
F2      8677
D1      7522
T1      7458
N1      7288
D2      7161
EC      7134
P1      6626
B1      6559
ARG     5330
USA     5233
SC0     5124
BRA     4850
G1      4128
SC1     4094
SC2     4092
JAP     4089
MEX     4080
SC3     4010
ROM     3640
POL     3567
SWE     3110
NOR     3092
RUS     3008
DEN     2634
CHN     2544
IRL     2367
FIN     2354
AUT     2306
SUI     2273
Name: count, dtype: int64

Filtro el dataset a primera y segunda división de las cinco grandes ligas 
(Inglaterra, España, Italia, Alemania y Francia) usando `isin()` sobre una lista de 
códigos. Verifico las dimensiones antes y después para confirmar el filtrado.

In [261]:
ligas_grandes = ["E0", "E1", "SP1", "SP2", "I1", "I2", "D1", "D2", "F1", "F2"]

df_matches_filtrado = df_matches[df_matches["Division"].isin(ligas_grandes)]

print("Antes del filtro:", df_matches.shape)
print("Después del filtro:", df_matches_filtrado.shape)

Antes del filtro: (230557, 48)
Después del filtro: (93570, 48)


In [262]:
resumen_nulos(df_matches_filtrado)

,nulos,porcentaje
MatchTime,70562,75.41
C_LTA,29918,31.97
C_VHD,29918,31.97
C_VAD,29918,31.97
C_LTH,29918,31.97
C_HTB,29918,31.97
C_PHB,29918,31.97
HomeTarget,29156,31.16
AwayTarget,29156,31.16
HomeFouls,29108,31.11


### Limpieza de columnas

Descarto columnas sin valor analítico: `MatchTime` (mayoritariamente vacía) y las 
columnas de apuestas especializadas. Conservo el núcleo de resultados, las estadísticas 
de partido y la cuota 1X2 básica.

In [263]:
columnas_a_soltar = [
    # Grupo 1: poco valor analítico
    "MatchTime",
    "C_LTH", "C_LTA", "C_VHD", "C_VAD", "C_HTB", "C_PHB",
    "MaxHome", "MaxDraw", "MaxAway",
    "Over25", "Under25", "MaxOver25", "MaxUnder25",
    "HandiSize", "HandiHome", "HandiAway",
]

df_matches_limpio = df_matches_filtrado.drop(columns=columnas_a_soltar)

print("Antes de soltar columnas:", df_matches_filtrado.shape)
print("Después de soltar columnas:", df_matches_limpio.shape)

Antes de soltar columnas: (93570, 48)
Después de soltar columnas: (93570, 31)


### Conversión de fechas

Convierto las columnas de fecha de texto a tipo `datetime`, lo que permite ordenar, 
filtrar y extraer componentes temporales, además de habilitar el cruce temporal con 
la tabla de Elo.

In [264]:
df_matches_limpio["MatchDate"] = pd.to_datetime(df_matches_limpio["MatchDate"])
df_elo["date"] = pd.to_datetime(df_elo["date"])

print(df_matches_limpio["MatchDate"].dtype)
print(df_elo["date"].dtype)

datetime64[us]
datetime64[us]


### Variables temporales derivadas

Extraigo el año y el mes de cada partido. A partir de la fecha, también asignaré la 
temporada futbolística a la que pertenece cada encuentro.

In [265]:
df_matches_limpio["anio"] = df_matches_limpio["MatchDate"].dt.year
df_matches_limpio["mes"] = df_matches_limpio["MatchDate"].dt.month

In [266]:
def asignar_temporada(fecha):
    """
    Asigna la temporada futbolística europea a una fecha dada.

    Una temporada va de agosto a mayo del año siguiente. Los partidos
    de agosto a diciembre pertenecen a la temporada que empieza ese año;
    los de enero a julio, a la temporada que empezó el año anterior.

    Parámetros:
    - fecha (datetime): fecha del partido.

    Retorna:
    - str: temporada en formato 'AAAA/AA' (por ejemplo, '2023/24').
    """
    anio = fecha.year
    mes = fecha.month

    if mes >= 8:
        anio_inicio = anio
    else:
        anio_inicio = anio - 1

    anio_fin = anio_inicio + 1
    return f"{anio_inicio}/{str(anio_fin)[-2:]}"


df_matches_limpio["temporada"] = df_matches_limpio["MatchDate"].apply(asignar_temporada)

In [267]:
df_matches_limpio[["MatchDate", "anio", "mes", "temporada"]].sample(10)

,MatchDate,anio,mes,temporada
66859,2011-02-12,2011,2,2010/11
18591,2003-11-22,2003,11,2003/04
1791,2000-11-11,2000,11,2000/01
68939,2011-04-25,2011,4,2010/11
60057,2010-02-20,2010,2,2009/10
16993,2003-08-25,2003,8,2003/04
120815,2016-02-02,2016,2,2015/16
107738,2015-01-09,2015,1,2014/15
158107,2019-02-22,2019,2,2018/19
72659,2011-12-04,2011,12,2011/12


## 3. Cruce con las puntuaciones Elo

El objetivo del cruce no es duplicar el Elo que ya traen los partidos, sino aportar 
valor: calcular el **Elo de cada club al inicio de la temporada** y usarlo como medida 
de expectativa previa. Antes de unir, verifico la tabla de Elo y la clave de cruce.

In [268]:
print("Rango de fechas:", df_elo["date"].min(), "->", df_elo["date"].max())
print("Clubes únicos:", df_elo["club"].nunique())
print("Países únicos:", df_elo["country"].nunique())
print("Fechas distintas:", df_elo["date"].nunique())

Rango de fechas: 2000-07-01 00:00:00 -> 2025-06-01 00:00:00
Clubes únicos: 895
Países únicos: 19
Fechas distintas: 599


### Verificación previa al cruce

Sigo las buenas prácticas antes de un merge: compruebo que no haya duplicados en la 
clave (club + fecha) y verifico el solapamiento de nombres de club entre ambas tablas 
mediante operaciones de conjuntos.

In [269]:
duplicados_elo = df_elo.duplicated(subset=["club", "date"]).sum()
print("Filas duplicadas por (club, date):", duplicados_elo)

Filas duplicadas por (club, date): 0


In [270]:
clubes_matches = set(df_matches_limpio["HomeTeam"]) | set(df_matches_limpio["AwayTeam"])
clubes_elo = set(df_elo["club"])

clubes_sin_elo = clubes_matches - clubes_elo

print("Clubes distintos en matches:", len(clubes_matches))
print("Clubes distintos en elo:", len(clubes_elo))
print("Clubes de matches que SÍ están en elo:", len(clubes_matches & clubes_elo))
print("Clubes de matches que NO están en elo:", len(clubes_sin_elo))

Clubes distintos en matches: 386
Clubes distintos en elo: 895
Clubes de matches que SÍ están en elo: 376
Clubes de matches que NO están en elo: 10


In [271]:
sorted(clubes_sin_elo)

['Ajaccio GFCO',
 'Andorra',
 'Extremadura UD',
 'FeralpiSalo',
 'Gimnastic',
 'Kaiserslautern ',
 'Piacenza ',
 'Preußen Münster',
 'Rouen',
 'Stuttgarter K']

Detecto que algunos clubes no cruzan por espacios sobrantes en el nombre. Los limpio 
con `str.strip()` en ambas tablas para normalizar la clave de cruce.

In [272]:
df_matches_limpio["HomeTeam"] = df_matches_limpio["HomeTeam"].str.strip()
df_matches_limpio["AwayTeam"] = df_matches_limpio["AwayTeam"].str.strip()
df_elo["club"] = df_elo["club"].str.strip()

In [273]:
clubes_matches = set(df_matches_limpio["HomeTeam"]) | set(df_matches_limpio["AwayTeam"])
clubes_elo = set(df_elo["club"])
clubes_sin_elo = clubes_matches - clubes_elo

print("Clubes que NO están en elo:", len(clubes_sin_elo))
sorted(clubes_sin_elo)

Clubes que NO están en elo: 8


['Ajaccio GFCO',
 'Andorra',
 'Extremadura UD',
 'FeralpiSalo',
 'Gimnastic',
 'Preußen Münster',
 'Rouen',
 'Stuttgarter K']

Cuantifico el impacto real de los clubes sin Elo: mido qué porcentaje de partidos se 
verá afectado y en qué divisiones se concentra, para decidir con criterio si conservarlos.

In [274]:
partidos_afectados = df_matches_limpio[
    df_matches_limpio["HomeTeam"].isin(clubes_sin_elo) |
    df_matches_limpio["AwayTeam"].isin(clubes_sin_elo)
]

n_afectados = len(partidos_afectados)
porcentaje = (n_afectados / len(df_matches_limpio)) * 100

print(f"Partidos afectados: {n_afectados} de {len(df_matches_limpio)} ({porcentaje:.2f}%)")

Partidos afectados: 983 de 93570 (1.05%)


In [275]:
partidos_afectados["Division"].value_counts()

Division
SP2    665
F2     191
D2      51
F1      38
I2      38
Name: count, dtype: int64

### Cálculo del Elo de inicio de temporada

Asigno la temporada a cada registro de Elo y, ordenando por fecha, me quedo con el 
primer registro de cada club en cada temporada mediante `drop_duplicates`. Así obtengo 
el Elo de arranque de cada equipo.

In [276]:
df_elo["temporada"] = df_elo["date"].apply(asignar_temporada)
print(df_elo.shape)
df_elo.head()

(245033, 5)


,date,club,country,elo,temporada
0,2000-07-01,Aachen,GER,1453.60,1999/00
1,2000-07-01,Aalborg,DEN,1482.61,1999/00
2,2000-07-01,Aalst,BEL,1337.53,1999/00
3,2000-07-01,Aarhus,DEN,1381.46,1999/00
4,2000-07-01,Aberdeen,SCO,1360.43,1999/00


In [277]:
df_elo_ordenado = df_elo.sort_values("date")

df_elo_inicio = df_elo_ordenado.drop_duplicates(subset=["club", "temporada"], keep="first")

print("Filas originales de elo:", len(df_elo))
print("Filas de elo inicio de temporada:", len(df_elo_inicio))

Filas originales de elo: 245033
Filas de elo inicio de temporada: 11788


In [278]:
elo_inicio_temporada = df_elo_inicio[["club", "temporada", "country", "elo"]]
elo_inicio_temporada = elo_inicio_temporada.rename(columns={"elo": "elo_inicio"})

print(elo_inicio_temporada.shape)
elo_inicio_temporada.head()

(11788, 4)


,club,temporada,country,elo_inicio
0,Aachen,1999/00,GER,1453.60
274,Piacenza,1999/00,ITA,1598.04
273,Petrolul Ploiesti,1999/00,ROM,1330.04
272,Pescara,1999/00,ITA,1563.65
271,Perugia,1999/00,ITA,1658.87


In [279]:
df_matches_limpio["temporada"].value_counts().sort_index()

temporada
1999/00      19
2000/01    3775
2001/02    3597
2002/03    2767
2003/04    2394
2004/05    3207
2005/06    3979
2006/07    3146
2007/08    3930
2008/09    3988
2009/10    3988
2010/11    4015
2011/12    3971
2012/13    3996
2013/14    3970
2014/15    4008
2015/16    3977
2016/17    3997
2017/18    3980
2018/19    3877
2019/20    3675
2020/21    3950
2021/22    3911
2022/23    3876
2023/24    3822
2024/25    3755
Name: count, dtype: int64

### Detección de un error: la "temporada fantasma"

Al revisar el reparto de partidos por temporada, detecto una anomalía: aparece una 
temporada **"1999/00" con solo 19 partidos**, frente a los ~3.900 de una temporada normal.

La causa es el mes de corte. Al usar agosto (`mes >= 8`), los partidos de finales de 
julio —que algunas ligas como la Ligue 1 francesa ya disputan— quedan asignados a la 
temporada anterior, creando una temporada "fantasma" que en realidad no existe.

**Corrección:** cambio el mes de corte a julio (`mes >= 7`), de modo que esos partidos 
se asignen a la temporada que realmente arranca. A continuación redefino la función 
corregida y vuelvo a aplicarla a ambas tablas.

In [280]:
def asignar_temporada(fecha):
    """
    Asigna la temporada futbolística europea a una fecha dada.

    Una temporada va de julio a mayo del año siguiente. Se toma julio como
    mes de corte porque algunas ligas (p. ej. la Ligue 1 francesa) disputan
    sus primeras jornadas a finales de julio. Los partidos de julio a
    diciembre pertenecen a la temporada que empieza ese año; los de enero
    a junio, a la temporada que empezó el año anterior.

    Parámetros:
    - fecha (datetime): fecha del partido o del registro de Elo.

    Retorna:
    - str: temporada en formato 'AAAA/AA' (por ejemplo, '2023/24').
    """
    anio = fecha.year
    mes = fecha.month

    if mes >= 7:
        anio_inicio = anio
    else:
        anio_inicio = anio - 1

    anio_fin = anio_inicio + 1
    return f"{anio_inicio}/{str(anio_fin)[-2:]}"

In [281]:
df_matches_limpio["temporada"] = df_matches_limpio["MatchDate"].apply(asignar_temporada)
df_elo["temporada"] = df_elo["date"].apply(asignar_temporada)

In [282]:
df_elo_ordenado = df_elo.sort_values("date")
df_elo_inicio = df_elo_ordenado.drop_duplicates(subset=["club", "temporada"], keep="first")

elo_inicio_temporada = df_elo_inicio[["club", "temporada", "country", "elo"]]
elo_inicio_temporada = elo_inicio_temporada.rename(columns={"elo": "elo_inicio"})

In [283]:
df_matches_limpio["temporada"].value_counts().sort_index()

temporada
2000/01    3758
2001/02    3633
2002/03    2767
2003/04    2394
2004/05    3188
2005/06    3988
2006/07    3146
2007/08    3940
2008/09    3988
2009/10    3988
2010/11    3988
2011/12    3988
2012/13    3988
2013/14    3988
2014/15    3988
2015/16    3988
2016/17    3988
2017/18    3988
2018/19    3868
2019/20    3273
2020/21    4338
2021/22    3906
2022/23    3905
2023/24    3831
2024/25    3755
Name: count, dtype: int64

In [284]:
print("Total de partidos:", df_matches_limpio["temporada"].value_counts().sum())
print("Filas del DataFrame:", len(df_matches_limpio))

Total de partidos: 93570
Filas del DataFrame: 93570


### Unión de las dos fuentes (merge doble)

Uno la tabla de partidos con el Elo de inicio de temporada. Como cada partido tiene 
dos equipos, hago **dos merges**: uno para el local y otro para el visitante, cruzando 
por club y temporada. Uso `how="left"` para conservar todos los partidos, aunque falte 
el Elo de algún equipo. Verifico que el número de filas no cambia tras cada unión.

In [285]:
df_final = df_matches_limpio.merge(
    elo_inicio_temporada,
    how="left",
    left_on=["HomeTeam", "temporada"],
    right_on=["club", "temporada"]
)

df_final = df_final.rename(columns={
    "elo_inicio": "elo_inicio_home",
    "country": "country_home"
})

df_final = df_final.drop(columns=["club"])

print("Antes del merge:", df_matches_limpio.shape)
print("Después del merge 1:", df_final.shape)

Antes del merge: (93570, 34)
Después del merge 1: (93570, 36)


In [286]:
df_final = df_final.merge(
    elo_inicio_temporada,
    how="left",
    left_on=["AwayTeam", "temporada"],
    right_on=["club", "temporada"]
)

df_final = df_final.rename(columns={
    "elo_inicio": "elo_inicio_away",
    "country": "country_away"
})

df_final = df_final.drop(columns=["club"])

print("Después del merge 2:", df_final.shape)

Después del merge 2: (93570, 38)


In [287]:
print("Nulos en elo_inicio_home:", df_final["elo_inicio_home"].isnull().sum())
print("Nulos en elo_inicio_away:", df_final["elo_inicio_away"].isnull().sum())

sin_elo = df_final["elo_inicio_home"].isnull() | df_final["elo_inicio_away"].isnull()
print(f"Partidos sin Elo en alguno de los dos: {sin_elo.sum()} ({sin_elo.mean()*100:.2f}%)")

Nulos en elo_inicio_home: 620
Nulos en elo_inicio_away: 629
Partidos sin Elo en alguno de los dos: 1247 (1.33%)


In [288]:
df_final[["MatchDate", "temporada", "HomeTeam", "AwayTeam", 
          "elo_inicio_home", "elo_inicio_away", 
          "country_home", "country_away"]].sample(10)

,MatchDate,temporada,HomeTeam,AwayTeam,elo_inicio_home,elo_inicio_away,country_home,country_away
14544,2005-02-02,2004/05,Messina,Milan,1552.09,1854.12,ITA,ITA
15365,2005-04-24,2004/05,Poli Ejido,Cadiz,1561.32,1593.85,ESP,ESP
11863,2004-03-18,2003/04,Ternana,Como,1541.16,1563.18,ITA,ITA
84399,2023-02-11,2022/23,Sevilla,Mallorca,1826.56,1615.09,ESP,ESP
46117,2013-04-13,2012/13,Fortuna Dusseldorf,Werder Bremen,1612.55,1679.34,GER,GER
73575,2020-06-14,2019/20,Cadiz,Vallecano,1563.32,1583.41,ESP,ESP
81473,2022-04-15,2021/22,Hull,Cardiff,1443.22,1568.33,ENG,ENG
89059,2024-04-03,2023/24,Arsenal,Luton,1919.38,1606.54,ENG,ENG
65298,2018-02-17,2017/18,Angers,St Etienne,1580.66,1624.58,FRA,FRA
71050,2019-09-15,2019/20,Lugo,Fuenlabrada,1474.67,1430.34,ESP,ESP


In [289]:
falta_elo_home = df_final["elo_inicio_home"].isnull() & df_final["HomeTeam"].isin(clubes_elo)
falta_elo_away = df_final["elo_inicio_away"].isnull() & df_final["AwayTeam"].isin(clubes_elo)

print("Partidos con local en ClubElo pero sin elo de esa temporada:", falta_elo_home.sum())
print("Partidos con visitante en ClubElo pero sin elo de esa temporada:", falta_elo_away.sum())

Partidos con local en ClubElo pero sin elo de esa temporada: 130
Partidos con visitante en ClubElo pero sin elo de esa temporada: 134


In [290]:
casos_temporada = df_final[falta_elo_home][["temporada", "HomeTeam"]].drop_duplicates()
print("Combinaciones club-temporada sin Elo:", len(casos_temporada))
casos_temporada.sort_values("temporada").head(20)

Combinaciones club-temporada sin Elo: 6


,temporada,HomeTeam
12577,2004/05,Plymouth
15817,2005/06,Plymouth
19763,2006/07,Plymouth
23010,2007/08,Plymouth
26850,2008/09,Plymouth
30882,2009/10,Plymouth


## 4. Variables derivadas para el análisis

Creo las variables que conectan la expectativa previa con el resultado real: la 
diferencia de Elo entre rivales (`diff_elo_inicio`), el total y la diferencia de goles, 
y los puntos obtenidos por el local. Estas variables alimentarán tanto el análisis 
estadístico como el dashboard.

In [291]:
df_final["diff_elo_inicio"] = df_final["elo_inicio_home"] - df_final["elo_inicio_away"]

df_final["total_goles"] = df_final["FTHome"] + df_final["FTAway"]
df_final["diff_goles"] = df_final["FTHome"] - df_final["FTAway"]

def puntos_local(resultado):
    """
    Traduce el resultado de un partido a los puntos que obtuvo el equipo local.

    Sigue el sistema estándar del fútbol: 3 puntos por victoria,
    1 por empate, 0 por derrota.

    Parámetros:
    - resultado (str): 'H' (gana local), 'D' (empate), 'A' (gana visitante).

    Retorna:
    - int: puntos obtenidos por el equipo local (3, 1 o 0).
    """
    if resultado == "H":
        return 3
    elif resultado == "D":
        return 1
    else:
        return 0


df_final["puntos_local"] = df_final["FTResult"].apply(puntos_local)

print(df_final.shape)
df_final[["HomeTeam", "AwayTeam", "FTHome", "FTAway", "FTResult",
          "diff_elo_inicio", "total_goles", "diff_goles", "puntos_local"]].sample(8)

(93570, 42)


,HomeTeam,AwayTeam,FTHome,FTAway,FTResult,diff_elo_inicio,total_goles,diff_goles,puntos_local
26647,Osasuna,Murcia,2.0,1.0,H,147.92,3.0,1.0,3
9232,Metz,Istres,4.0,1.0,H,175.28,5.0,3.0,3
7296,Stuttgart,Kaiserslautern,4.0,3.0,H,-25.16,7.0,1.0,3
32083,Mantova,Cesena,0.0,1.0,A,78.70,1.0,-1.0,0
92282,Huesca,Granada,1.0,1.0,D,-40.83,2.0,0.0,1
40636,Sochaux,Ajaccio,0.0,2.0,A,162.77,2.0,-2.0,0
90950,Amiens,Martigues,1.0,1.0,D,56.57,2.0,0.0,1
57578,Virtus Entella,Bari,2.0,0.0,H,-69.76,2.0,2.0,3


## 5. Análisis descriptivo y estadístico

Con el dataset final construido, exploro los patrones principales y los contrasto con 
pruebas de hipótesis. Cada test sigue la misma estructura: planteamiento (hipótesis y 
nivel de significación), ejecución e interpretación crítica de los resultados.

In [292]:
print("Distribución de resultados (%):")
print(df_final["FTResult"].value_counts(normalize=True) * 100)

print("\nMedia de goles del local:", df_final["FTHome"].mean().round(3))
print("Media de goles del visitante:", df_final["FTAway"].mean().round(3))
print("Media de goles por partido:", df_final["total_goles"].mean().round(3))

Distribución de resultados (%):
FTResult
H    44.831677
D    27.727904
A    27.440419
Name: proportion, dtype: float64

Media de goles del local: 1.469
Media de goles del visitante: 1.111
Media de goles por partido: 2.58


### Test de hipótesis: ¿existe la ventaja de jugar en casa?

El análisis descriptivo muestra que el local marca más goles de media (1,469) 
que el visitante (1,111). Para comprobar si esta diferencia es real y no fruto 
del azar, aplico una **prueba t de Student para muestras pareadas** (`ttest_rel`).

Uso la versión pareada porque los goles del local y del visitante proceden del 
mismo partido, es decir, son observaciones emparejadas, no dos grupos independientes.

- **H₀:** no hay diferencia entre la media de goles del local y del visitante.
- **H₁:** la media de goles del local es distinta a la del visitante.
- **Nivel de significación:** α = 0,05.

In [293]:
from scipy import stats

goles_local = df_final["FTHome"]
goles_visitante = df_final["FTAway"]

print("Nulos en goles local:", goles_local.isnull().sum())
print("Nulos en goles visitante:", goles_visitante.isnull().sum())

t_stat, p_valor = stats.ttest_rel(goles_local, goles_visitante)

print(f"\nEstadístico t: {t_stat:.4f}")
print(f"p-valor: {p_valor}")

Nulos en goles local: 0
Nulos en goles visitante: 0

Estadístico t: 66.0533
p-valor: 0.0


### Interpretación

El estadístico t es 66,05 y el p-valor es prácticamente cero (**p < 0,001**), 
muy por debajo de α = 0,05. Por tanto, **rechazo la hipótesis nula**.

Existe evidencia estadística muy fuerte de que el equipo local marca más goles 
que el visitante: la ventaja de jugar en casa es un efecto real, no azar.

**Matiz sobre el tamaño del efecto:** con una muestra tan grande (93.570 partidos), 
el test detecta como significativa casi cualquier diferencia. Por eso, más allá 
del p-valor, conviene mirar la magnitud real: la diferencia es de ~0,36 goles por 
partido a favor del local, que en el contexto del fútbol sí es relevante.

In [294]:
from scipy import stats

datos_validos = df_final.dropna(subset=["diff_elo_inicio", "diff_goles"])

print("Partidos usados en el test:", len(datos_validos))
print("Partidos descartados por falta de Elo:", len(df_final) - len(datos_validos))

coef_pearson, p_valor = stats.pearsonr(
    datos_validos["diff_elo_inicio"],
    datos_validos["diff_goles"]
)

print(f"\nCoeficiente de correlación (r): {coef_pearson:.4f}")
print(f"p-valor: {p_valor}")

Partidos usados en el test: 92323
Partidos descartados por falta de Elo: 1247

Coeficiente de correlación (r): 0.3022
p-valor: 0.0


### Interpretación

La correlación de Pearson entre la ventaja de Elo inicial del local 
(`diff_elo_inicio`) y su diferencia de goles (`diff_goles`) es r = 0,30 
(p < 0,001), calculada sobre los 92.323 partidos con Elo disponible 
(se descartan 1.247 partidos, el 1,33%, por no tener Elo de referencia).

Se trata de una correlación positiva y moderada: cuanto mayor es la ventaja 
de Elo del equipo local al inicio de la temporada, mayor tiende a ser su 
diferencia de goles a favor. Esto confirma que el Elo previo aporta información 
real sobre el resultado y justifica la unión de ambas fuentes de datos.

Sobre la magnitud: un valor de 0,30 es el esperable en fútbol, un deporte de 
alta varianza donde el azar (penaltis, rebotes, expulsiones) pesa mucho en cada 
partido. El Elo capta la calidad estructural del equipo, pero no puede predecir 
lo que ocurre en 90 minutos concretos. Una correlación moderada refleja justamente 
eso: una tendencia clara, pero con amplio margen para la sorpresa.

Matices metodológicos:
- La significación es extrema (p < 0,001) por el gran tamaño de muestra, por lo 
  que el dato informativo es la magnitud del coeficiente (0,30), no el p-valor.
- Correlación no implica causalidad: el Elo no causa los goles; ambos son reflejo 
  de la calidad subyacente del equipo.

In [295]:
goles_por_liga = df_final.groupby("Division")["total_goles"].mean().sort_values(ascending=False)
print(goles_por_liga.round(3))

Division
D1     2.948
D2     2.788
E0     2.718
I1     2.679
SP1    2.673
E1     2.564
F1     2.490
I2     2.418
F2     2.345
SP2    2.330
Name: total_goles, dtype: float64


### Test de hipótesis: ¿difieren los goles entre ligas?

Comparo la media de goles totales por partido entre las 10 ligas mediante un 
**ANOVA de una vía** (`f_oneway`), la prueba adecuada para comparar las medias 
de más de dos grupos en una sola contrastación.

- **H₀:** todas las ligas tienen la misma media de goles por partido.
- **H₁:** al menos una liga tiene una media distinta.
- **Nivel de significación:** α = 0,05.

In [296]:
from scipy import stats

grupos_goles = [
    grupo["total_goles"].values
    for nombre_liga, grupo in df_final.groupby("Division")
]

print("Número de grupos (ligas):", len(grupos_goles))

f_stat, p_valor = stats.f_oneway(*grupos_goles)

print(f"\nEstadístico F: {f_stat:.4f}")
print(f"p-valor: {p_valor}")

Número de grupos (ligas): 10

Estadístico F: 132.2706
p-valor: 5.756945186853497e-249


### Interpretación

El estadístico F es 132,27 con un p-valor prácticamente nulo (**p < 0,001**), 
por lo que **rechazo la hipótesis nula**: existen diferencias significativas en 
el número de goles entre ligas.

La tabla de medias revela el patrón concreto:
- La **Bundesliga (D1)** es la más goleadora (2,95 goles/partido), seguida de la 
  2. Bundesliga (2,79): el fútbol alemán destaca por su carácter ofensivo.
- Premier, Serie A y LaLiga se agrupan en torno a 2,67–2,72.
- Las segundas divisiones española (2,33) y francesa (2,35) son las menos goleadoras.

La diferencia entre la liga más y la menos goleadora supera el medio gol por 
partido, una brecha relevante en términos futbolísticos. El ANOVA confirma que 
estas diferencias no se deben al azar.

In [297]:
tabla_contingencia = pd.crosstab(df_final["FTResult"], df_final["Division"])
print(tabla_contingencia)

Division    D1    D2    E0    E1    F1    F2    I1    I2   SP1   SP2
FTResult                                                            
A         2231  2013  2782  3943  2378  2183  2551  2462  2495  2638
D         1857  1934  2318  3706  2435  2637  2439  3175  2271  3173
H         3434  3214  4310  5957  3943  3857  4022  4284  4242  4686


In [298]:
tabla_porcentajes = pd.crosstab(
    df_final["FTResult"], df_final["Division"], normalize="columns"
) * 100
print(tabla_porcentajes.round(1))

Division    D1    D2    E0    E1    F1    F2    I1    I2   SP1   SP2
FTResult                                                            
A         29.7  28.1  29.6  29.0  27.2  25.2  28.3  24.8  27.7  25.1
D         24.7  27.0  24.6  27.2  27.8  30.4  27.1  32.0  25.2  30.2
H         45.7  44.9  45.8  43.8  45.0  44.5  44.6  43.2  47.1  44.6


### Test de hipótesis: ¿depende el resultado de la liga?

Analizo si el resultado del partido (victoria local, empate o victoria visitante) 
está asociado a la liga, mediante una **prueba chi-cuadrado de independencia** 
(`chi2_contingency`) sobre la tabla de contingencia resultado × liga.

- **H₀:** el resultado y la liga son independientes (mismo reparto H/D/A en todas las ligas).
- **H₁:** el resultado y la liga están asociados.
- **Nivel de significación:** α = 0,05.

In [299]:
from scipy import stats

chi2, p_valor, grados_libertad, frecuencias_esperadas = stats.chi2_contingency(tabla_contingencia)

print(f"Estadístico chi-cuadrado: {chi2:.4f}")
print(f"p-valor: {p_valor}")
print(f"Grados de libertad: {grados_libertad}")

Estadístico chi-cuadrado: 323.2289
p-valor: 7.869562861490078e-58
Grados de libertad: 18


### Interpretación

El estadístico chi-cuadrado es 323,23 (18 grados de libertad) con un p-valor 
prácticamente nulo (**p < 0,001**), por lo que **rechazo la hipótesis nula**: 
existe asociación estadística entre el resultado y la liga.

Ahora bien, la magnitud de esa asociación debe matizarse observando la tabla de 
porcentajes:
- La **victoria local es muy estable** entre ligas: oscila solo entre el 43,2% (I2) 
  y el 47,1% (SP1). La ventaja de jugar en casa es un fenómeno universal del fútbol 
  europeo, no dependiente de la liga.
- Las diferencias significativas se concentran en el **reparto empate / victoria 
  visitante**. Las segundas divisiones italiana (I2) y francesa (F2) destacan por su 
  alto porcentaje de empates (32,0% y 30,4%).

Este resultado es coherente con el ANOVA de goles: las mismas ligas menos goleadoras 
(I2, F2) son las que más empates registran, lo que refuerza su perfil de competiciones 
más cerradas y tácticas. Como en los tests anteriores, la significación extrema deriva 
del gran tamaño muestral, por lo que la lectura relevante es la magnitud y el patrón, 
no el p-valor.

In [300]:
ruta_processed = "../data/processed/"

df_final.to_csv(ruta_processed + "matches_elo_final.csv", index=False)

print("Dataset final guardado.")
print("Dimensiones:", df_final.shape)

Dataset final guardado.
Dimensiones: (93570, 42)


In [301]:
df_verificacion = pd.read_csv(ruta_processed + "matches_elo_final.csv")

print("Dimensiones al releer:", df_verificacion.shape)
print("\nColumnas:")
print(df_verificacion.columns.tolist())
print("\nTipos de las columnas clave:")
print(df_verificacion[["MatchDate", "diff_elo_inicio", "total_goles", "puntos_local"]].dtypes)

Dimensiones al releer: (93570, 42)

Columnas:
['Division', 'MatchDate', 'HomeTeam', 'AwayTeam', 'HomeElo', 'AwayElo', 'Form3Home', 'Form5Home', 'Form3Away', 'Form5Away', 'FTHome', 'FTAway', 'FTResult', 'HTHome', 'HTAway', 'HTResult', 'HomeShots', 'AwayShots', 'HomeTarget', 'AwayTarget', 'HomeFouls', 'AwayFouls', 'HomeCorners', 'AwayCorners', 'HomeYellow', 'AwayYellow', 'HomeRed', 'AwayRed', 'OddHome', 'OddDraw', 'OddAway', 'anio', 'mes', 'temporada', 'country_home', 'elo_inicio_home', 'country_away', 'elo_inicio_away', 'diff_elo_inicio', 'total_goles', 'diff_goles', 'puntos_local']

Tipos de las columnas clave:
MatchDate              str
diff_elo_inicio    float64
total_goles        float64
puntos_local         int64
dtype: object
